java/ffi/src/main/java/sleeper/foreign/FFIArray.java

```diff
@@ -0,0 +1,277 @@
+        } else if (item instanceof Integer) {
+            int e = (int) item;
+            this.items[idx] = r.getMemoryManager().allocateDirect(r.findType(NativeType.SINT).size());
+            this.items[idx].putInt(0, e);
+        } else if (item instanceof Long) {
+            long e = (long) item;
+            this.items[idx] = r.getMemoryManager().allocateDirect(r.findType(NativeType.SLONGLONG).size());
+            this.items[idx].putLong(0, e);
+        } else if (item instanceof java.lang.String) {
+            // Strings are encoded as 4 byte length then value
+            java.lang.String e = (java.lang.String) item;
+            byte[] utf8string = e.getBytes(StandardCharsets.UTF_8);
+            // Add four for length
+            int stringSize = utf8string.length + 4;
+            // Allocate memory for string and write length then the string
+            this.items[idx] = r.getMemoryManager().allocateDirect(stringSize);
+            this.items[idx].putInt(0, utf8string.length);
+            this.items[idx].put(4, utf8string, 0, utf8string.length);
+        } else if (item instanceof byte[]) {
+            // Byte arrays are encoded as 4 byte length then value
+            byte[] e = (byte[]) item;
+            int byteSize = e.length + 4;
+            this.items[idx] = r.getMemoryManager().allocateDirect(byteSize);
+            this.items[idx].putInt(0, e.length);
+            this.items[idx].put(4, e, 0, e.length);
+        } else if (item instanceof Boolean) {
+            boolean e = (boolean) item;
+            this.items[idx] = r.getMemoryManager().allocateDirect(1);
+            this.items[idx].putByte(0, e ? (byte) 1 : (byte) 0);
+        } else {
+            throw new ClassCastException("Can't cast " + item.getClass() + " to a valid Sleeper row key type");
+        }
+    }
+
+    /**
+     * Reads a value from array.
+     *
+     * The value of the array element is read from previously allocated memory. The
+     * array MUST have been
+     * previously populated with {@link FFIArray#populate(Object[], boolean)}. The type
+     * of the array item
+     * to read is given in the {@code clazz} argument. This should match the generic
+     * type given.
+     *
+     * If nulls are not allowed in this array, but a null pointer is found, then an
+     * exception is thrown.
+     *
+     * @param  <E>                       the generic type of elements in the array
+     * @param  idx                       the index to read
+     * @param  clazz                     the class type of generic parameter E
+     * @param  nullsAllowed              if nulls may be present in this array
+     * @param  r                         the runtime this array was allocated with
+     * @return                           the array element, or {@code null} if nulls are allowed and the
+     *                                   pointer at
+     *                                   {@code idx} is 0.
```

Comment:
This has formatted oddly, can it be tidied please?

This can either be capitalised as a sentence or the full stop can be removed.

See the Oracle style guide under `@param`: "The description begins with a lowercase letter if it is a phrase (contains no verb), or an uppercase letter if it is a sentence. End the phrase with a period only if another phrase or sentence follows it."

https://www.oracle.com/uk/technical-resources/articles/java/javadoc-tool.html